File: src/jvm/package/src/doc-examples/java/questpdf/docexamples/codepatterns/CodePatternComponentProgressbarComponentExample.java
package questpdf.docexamples.codepatterns;

import com.questpdf.elements.DynamicContext;
import com.questpdf.fluent.Document;
import com.questpdf.helpers.Colors;
import com.questpdf.helpers.PageSizes;
import com.questpdf.infrastructure.DynamicComponentComposeResult;
import com.questpdf.infrastructure.IDynamicComponent;
import org.junit.jupiter.api.Test;
import questpdf.docexamples.DocExample;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.concurrent.ThreadLocalRandom;

public class CodePatternComponentProgressbarComponentExample extends DocExample {

    @Test
    public void example() throws IOException {
        var content = generateReport();
        Files.write(Path.of(output("code-pattern-dynamic-component-progressbar.pdf")), content);
    }

    public byte[] generateReport() {
        return Document
            .create(document -> {
                document.page(page -> {
                    page.size(PageSizes.getA4());
                    page.margin(50f);
                    page.defaultTextStyle(style -> style.fontSize(20f));

                    page.header().column(column -> {
                        column.item()
                            .text("MyBrick Set")
                            .fontSize(48f).fontColor(Colors.Blue.getDarken2()).bold();

                        column.item()
                            .text("Building Instruction")
                            .fontSize(24f);

                        column.item().height(15f);

                        column.item().dynamic(new PageProgressbarComponent());
                    });

                    page.content().paddingVertical(25f).column(column -> {
                        column.spacing(25f);

                        for (var i = 1; i <= 30; i++) {
                            column.item()
                                .background(Colors.Grey.getLighten3())
                                .height(ThreadLocalRandom.current().nextInt(4, 8) * 25)
                                .alignCenter()
                                .alignMiddle()
                                .text("Step " + i);
                        }
                    });

                    page.footer().dynamic(new PageNumberSideComponent());
                });
            })
            .generatePdf();
    }

    public static class PageProgressbarComponent implements IDynamicComponent {
        @Override
        public DynamicComponentComposeResult compose(DynamicContext context) {
            var content = context.createElement(element -> {
                var width = context.getAvailableSize().getWidth() * context.getPageNumber() / context.getTotalPages();

                element
                    .background(Colors.Blue.getLighten3())
                    .height(5f)
                    .width(width)
                    .background(Colors.Blue.getDarken2());
            });

            var result = new DynamicComponentComposeResult();
            result.setContent(content);
            result.setHasMoreContent(false);
            return result;
        }
    }

    public static class PageNumberSideComponent implements IDynamicComponent {
        @Override
        public DynamicComponentComposeResult compose(DynamicContext context) {
            var content = context.createElement(element -> {
                // The chainable Element(Func<IContainer, IContainer>) overload is not bridged;
                // the conditional alignment is applied inline instead.
                var aligned = context.getPageNumber() % 2 == 0 ? element.alignRight() : element.alignLeft();

                aligned
                    .text(text -> {
                        text.span("Page ");
                        text.currentPageNumber();
                    });
            });

            var result = new DynamicComponentComposeResult();
            result.setContent(content);
            result.setHasMoreContent(false);
            return result;
        }
    }
}

File: src/jvm/package/src/doc-examples/kotlin/questpdf/docexamples/codepatterns/CodePatternComponentProgressbarComponentExample.kt
package questpdf.docexamples.codepatterns

import com.questpdf.elements.DynamicContext
import com.questpdf.fluent.Document
import com.questpdf.helpers.Colors
import com.questpdf.helpers.PageSizes
import com.questpdf.infrastructure.DynamicComponentComposeResult
import com.questpdf.infrastructure.IDynamicComponent
import org.junit.jupiter.api.Test
import questpdf.docexamples.DocExample
import java.io.File
import kotlin.random.Random

class CodePatternComponentProgressbarComponentExample : DocExample() {

    @Test
    fun example() {
        val content = generateReport()
        File(output("code-pattern-dynamic-component-progressbar.pdf")).writeBytes(content)
    }

    fun generateReport(): ByteArray {
        return Document
            .create {
                page {
                    size(PageSizes.A4)
                    margin(50f)
                    defaultTextStyle { fontSize(20f) }

                    header().column {
                        item()
                            .text("MyBrick Set")
                            .fontSize(48f).fontColor(Colors.Blue.Darken2).bold()

                        item()
                            .text("Building Instruction")
                            .fontSize(24f)

                        item().height(15f)

                        item().dynamic(PageProgressbarComponent())
                    }

                    content().paddingVertical(25f).column {
                        spacing(25f)

                        for (i in 1..30) {
                            item()
                                .background(Colors.Grey.Lighten3)
                                .height((Random.nextInt(4, 8) * 25).toFloat())
                                .alignCenter()
                                .alignMiddle()
                                .text("Step $i")
                        }
                    }

                    footer().dynamic(PageNumberSideComponent())
                }
            }
            .generatePdf()
    }

    class PageProgressbarComponent : IDynamicComponent {
        override fun compose(context: DynamicContext): DynamicComponentComposeResult {
            val content = context.createElement {
                val width = context.availableSize.width * context.pageNumber / context.totalPages

                background(Colors.Blue.Lighten3)
                    .height(5f)
                    .width(width)
                    .background(Colors.Blue.Darken2)
            }

            return DynamicComponentComposeResult().apply {
                this.content = content
                hasMoreContent = false
            }
        }
    }

    class PageNumberSideComponent : IDynamicComponent {
        override fun compose(context: DynamicContext): DynamicComponentComposeResult {
            val content = context.createElement {
                // The chainable Element(Func<IContainer, IContainer>) overload is not bridged;
                // the conditional alignment is applied inline instead.
                val aligned = if (context.pageNumber % 2 == 0) alignRight() else alignLeft()

                aligned
                    .text {
                        span("Page ")
                        currentPageNumber()
                    }
            }

            return DynamicComponentComposeResult().apply {
                this.content = content
                hasMoreContent = false
            }
        }
    }
}
